src/Init/Core.lean

```diff
@@ -1680,40 +1680,64 @@ So, you are mainly losing the capability of type checking your development using
+`Associative op` says that `op` is an associative operation,
+i.e. `(a ∘ b) ∘ c = a ∘ (b ∘ c)`.
 -/
-class IsAssociative {α : Sort u} (op : α → α → α) where
+class Associative (op : α → α → α) : Prop where
   /-- An associative operation satisfies `(a ∘ b) ∘ c = a ∘ (b ∘ c)`. -/
   assoc : (a b c : α) → op (op a b) c = op a (op b c)
 
 /--
-`IsCommutative op` says that `op` is a commutative operation,
-i.e. `a ∘ b = b ∘ a`. It is used by the `ac_rfl` tactic.
+`Commutative op` says that `op` is a commutative operation,
+i.e. `a ∘ b = b ∘ a`.
 -/
-class IsCommutative {α : Sort u} (op : α → α → α) where
+class Commutative (op : α → α → α) : Prop where
   /-- A commutative operation satisfies `a ∘ b = b ∘ a`. -/
   comm : (a b : α) → op a b = op b a
 
 /--
-`IsIdempotent op` says that `op` is an idempotent operation,
-i.e. `a ∘ a = a`. It is used by the `ac_rfl` tactic
-(which also simplifies up to idempotence when available).
+`Idempotent op` says that `op` is an idempotent operation,
+i.e. `a ∘ a = a`.
 -/
-class IsIdempotent {α : Sort u} (op : α → α → α) where
+class Idempotent (op : α → α → α) : Prop where
   /-- An idempotent operation satisfies `a ∘ a = a`. -/
   idempotent : (x : α) → op x x = x
 
 /--
-`IsNeutral op e` says that `e` is a neutral operation for `op`,
-i.e. `a ∘ e = a = e ∘ a`. It is used by the `ac_rfl` tactic
-(which also simplifies neutral elements when available).
+A binary operation with an associated identity element that can be
+inferred through class inference.
+
+This intentionally does not have associated lemmas so it can be
+implemented on operations not intended for reasoning about.  See
+@LawfulIdentity@ for an extension with lemmas.
+ -/
+class HasIdentity (op : α → α → α) (o : outParam α) : Prop where
+
+/-- A binary operation with a left identity. -/
+class LeftIdentity (op : α → α → α) (o : outParam α) extends HasIdentity op o : Prop where
+  /-- Left identify -/
+  left_id : ∀ a, op o a = a
+
+/-- A binary operation with a right identity. -/
+class RightIdentity (op : α → α → α) (o : outParam α) extends HasIdentity op o : Prop where
+  /-- Right identify -/
+  right_id : ∀ a, op a o = a
+
+/-- A binary operation with a left and right identity. -/
+class LawfulIdentity (op : α → α → α) (o : outParam α) extends LeftIdentity op o, RightIdentity op o : Prop
+
+/--
+Class that simplifies instances of LawfulIdentity on commutative
+functions by requiring only left or right identity.
 -/
-class IsNeutral {α : Sort u} (op : α → α → α) (neutral : α) where
-  /-- A neutral element can be cancelled on the left: `e ∘ a = a`. -/
-  left_neutral : (a : α) → op neutral a = a
-  /-- A neutral element can be cancelled on the right: `a ∘ e = a`. -/
-  right_neutral : (a : α) → op a neutral = a
+class LawfulCommIdentity (op : α → α → α) (o : outParam α) [hc : Commutative op] extends LawfulIdentity op o : Prop where
+  left_id a := Eq.trans (hc.comm o a) (right_id a)
+  right_id a := Eq.trans (hc.comm a o) (left_id a)
```

Comment:
I have a feeling that this class will cause some "impedance mismatch" problems. The issue is that you have to implement it specially if you want to use theorems that refer to `[LawfulCommIdentity op o]`. Maybe instead, we can have instances that derive the lawful identity class from commutativity and either LeftIdentity or RightIdentity?

```lean
/-- A binary operation with a left identity. -/
class LeftIdentity (op : α → α → α) (o : outParam α) : Prop where
  /-- Left identify -/
  left_id : ∀ a, op o a = a

/-- A binary operation with a right identity. -/
class RightIdentity (op : α → α → α) (o : outParam α) : Prop where
  /-- Right identify -/
  right_id : ∀ a, op a o = a

/-- A binary operation with a left and right identity. -/
class Identity (op : α → α → α) (o : outParam α) extends LeftIdentity op o, RightIdentity op o : Prop

instance (op : α → α → α) [LeftIdentity op o] [Commutative op] : Identity op o where
  right_id a := (Commutative.comm a o).trans (LeftIdentity.left_id a)

instance (op : α → α → α) [RightIdentity op o] [Commutative op] : Identity op o where
  left_id a := (Commutative.comm o a).trans (RightIdentity.right_id a)
```
That way theorems that need two-sided identities and a commutative operation can require `[Commutative op] [Identity op o]` and apply more generally than `[Commutative op] [LawfulCommIdentity op o]`.

We could consider making these be lower-priority instances, to keep them from being explored unnecessarily.

Reply:
I would also be tempted to have 
```lean
class abbrev Identity (op : α → α → α) (o : outParam α) := LeftIdentity op o, RightIdentity op o
```
and perhaps even
```lean
class abbrev Monoid (op : α → α → α) (o : outParam α) := Associative op, Identity op o
```
I think these classes are intended to have very few basic instances, so I don't think these are problematic.

Reply:
I just pushed some changes to Identity from the other thread.  I'm currently keeping `LawfulCommIdentity` but added a comment that it's not intended for constraints, but only instances.

I don't have a major objection to @kmill's alternative though.  Presumedly there's cycle detection to rule out search loops like searching for `LeftIdentity` to find `Identity` using commutative instance, then searching for `Identity` to find `LeftIdentity` using the extends clause.